feat(auth): oauth revoke endpoint

mdtro opened the pull request:
Implementation of the `/revoke` OAuth endpoint defined in [RFC 7009](https://www.rfc-editor.org/rfc/rfc7009) to allow confidential clients (ie. API Applications, Sentry Integrations, and Sentry Apps) to revoke access and refresh tokens.

mdtro committed:
wip: initial implementation of a revoke endpoint for oauth tokens

misha-sentry commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,127 @@
+    def error(self, request: HttpRequest, name, reason=None, status=400):
+        client_id = request.POST.get("client_id")
+
+        logging.error(
```
Can add any other additional fields to logging here, such as the user who made the request?

Are we also able to add logging in general when a request comes in and doesn't result in a error?

mdtro replied:
It would be the client performing the revoke and not the user. What information would be most helpful here to log for your team?

I was thinking `client_id` and a hashed version of the `token`.

misha-sentry replied:
I saw you already checked https://www.notion.so/sentry/Logging-Strategy-and-Guide-41d031bf2caf457d8eec79a8488102a5?pvs=4, but generally being able to answer who, what, when, and from where (IP or other location identifying attributes) is helpful. 

client_id looks like it's already in there. Hashed version of the `token` would be useful. 
Would it good to have a hashed version of the `refresh_token` if the `token` you are referring to is the `access_token`? 
what else can the client_id be tied to that could be useful (ie. application_id, org related info, owner of the token)?

mdtro replied:
Added the additional fields. :)

mdtro committed:
fix testcase import

mdtro committed:
add tests and align errors with OAuth 2.0 RFCs

- add additional test coverage for error cases
- align error responses with RFC 7009 Section 4.1.1
  and RFC 6749 Section 5.2
  - https://datatracker.ietf.org/doc/html/rfc7009#section-4.1.1
  - https://datatracker.ietf.org/doc/html/rfc6749#section-5.2

mdtro committed:
set reason to error_description to match oauth rfcs

mdtro committed:
update client auth error to match rfc and add more tests

mdtro committed:
more tests

mdtro committed:
add info log for successful revoke

mdtro committed:
fix encoding

mdtro committed:
fix typing error

mdtro committed:
add application_id to log data

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+
+    Clients can provide either the access_token or refresh_token in the request.
+
+    When revoking the token, the associated access_token or refresh_token is also
```
Does this mean that if I pass in an `access_token`, then any `refresh_token` attached to it would be revoked as well (and vice versa)?

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+    def dispatch(self, request, *args, **kwargs):
+        return super().dispatch(request, *args, **kwargs)
+
+    def error(self, request: HttpRequest, name, error_description=None, status=400):
```
Can we add typing to this (and the rest of the methods in here)?

mdtro replied:
Hmm. Added some typing here, but I'm running into this error in the CI. Have any advice on the proper way to fix this?

> src/sentry/web/frontend/oauth_revoke.py:131: error: Item "None" of "Optional[str]" has no attribute "encode"  [union-attr]

```python
        # only delete the token if one was found
        if token_to_delete:
            token_to_delete.delete()
            logger.info(
                "oauth.revoke-success",
                extra={
                    "client_id": client_id,
                    "application_id": application.id,
                    # don't log the actual token, just a hash of it
                    "sha256_provided_token": sha256(token.encode("utf-8")).hexdigest(),
                    "sha256_access_token": sha256(
                        token_to_delete.token.encode("utf-8")
                    ).hexdigest(),
                    "sha256_refresh_token": sha256(
                        token_to_delete.refresh_token.encode("utf-8")
                    ).hexdigest(),
                    "resource_owner_id": token_to_delete.user.id,
                },
            )
```

I have the `if token_to_delete:` in place, but this doesn't seem to be enough to satisfy mypy. I've also tried this with `if isinstance(token_to_delete, ApiToken):` with the same result.

mdtro replied:
Ah, actually I should have paid closer attention. It is not typing error on `ApiToken`, but rather it's `refresh_token` property that can be `None`.

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+    @method_decorator(never_cache)
+    def post(self, request: HttpRequest) -> HttpResponse:
+        """
+        Handles POST request to revoke an access_token or refresh_token.
```
```suggestion
        Revokes an access_token or refresh_token.
```

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+
+    def _get_token_to_delete(
+        self, token: str, token_type_hint: Optional[str], application: ApiApplication
+    ) -> Optional[ApiToken]:
```
nit: we're moving away from `Optional[..]` towards something like `ApiToken | None`. Will be fixed automatically at some point, but just fyi

mdtro replied:
Done. :)

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+            elif token_type_hint == "refresh_token":
+                token_to_delete = ApiToken.objects.get(refresh_token=token, application=application)
+            else:
+                # the client request did not provide a token hint so we must check both token (aka. access_token)
```
```suggestion
                # the client request did not provide a token hint so we must check both token types (aka. access_token)
```

EricHasegawa commented on src/sentry/web/frontend/oauth_revoke.py:
```diff
@@ -0,0 +1,155 @@
+                status=401,
+            )
+
+        token_to_delete: ApiToken = self._get_token_to_delete(
```
Incorrect typing I think - this could also be `None` right?

EricHasegawa commented on tests/sentry/web/frontend/test_oauth_revoke.py:
```diff
@@ -0,0 +1,245 @@
+        # create some oauth tokens
+        self.tokens: List[ApiToken] = []
+
+        for i in range(4):
```
nit: you could condense these two loops into one (would use fewer lines of code and be just a tiny bit faster)

mdtro replied:
Good idea! Thank you!

EricHasegawa commented on tests/sentry/web/frontend/test_oauth_revoke.py:
```diff
@@ -0,0 +1,245 @@
+            },
+        )
+
+        # seems odd at first, but per the RFC (https://www.rfc-editor.org/rfc/rfc7009#section-2.2)
```
This is so bad actors get really limited information from this endpoint right? Like they can't really check if they own a token or not?

mdtro replied:
Correct. It prevents enumeration attacks.

EricHasegawa reviewed (commented):
This looks great :) Left some minor comments/questions. Thanks for doing this!

mdtro committed:
address pr feedback

mdtro committed:
add backticks on `token` in comment to clarify it is referencing a column

mdtro committed:
add httprequest type on dispatch function

mdtro committed:
fix typing error on token_to_delete

mdtro committed:
fix type error on option token_to_delete.refresh_token

mdtro committed:
update post(..) method docstring

EricHasegawa reviewed (approved):
Looks great :) I'd recommend merging after hackweek just in case!

misha-sentry reviewed (changes requested):
We had an offline discussion about adding in a few more fields for logging, just going to set the review to request changes for that reason

misha-sentry commented:
**Currently:**
"error_name": name,
"status": status,
"client_id": client_id,
"reason": error_description,

**Proposed:** (if these fields are possible)
"error_name": name (system generated error code),
"status": status (status of the token revoking action),
 "client_id": client_id (client_id of the app),
"app_id": application_id (name of the applications associated with the client id)
"reason": error_description
"token_type": token_type (the type of token that was requested to be revoked, either access or refresh)
"resource_owner_id": resource_owner_id (either the user that authorized the 3rd-party app or the application owner or the user may be both)
"hashed_token": hashed_token (this will be a hash of the token, although we store our tokens in plain-text, is there a way to map back which token it is?)

Context on resource_owner_id: 
If app owner != User_id,
Then resource_owner_id = user that authorized third-party app.

If app owner == user_id,
Then resource_owner_id = user that both owns the app and is also the only user involved in this event

And resource_owner_id will never be null, it will always be the user that the app was authorized on. Regardless of whether they actually own the app or it’s a third-party app they authorized

-------------------------------
For Matt - You mentioned "
"sha256_provided_token": sha256_provided_token,
"sha256_access_token": sha256_access_token,
"sha256_refresh_token": sha256_refresh_token,"

When would the provided token be different than the hashed access/refresh token?

mdtro commented:
> For Matt - You mentioned "
"sha256_provided_token": sha256_provided_token,
"sha256_access_token": sha256_access_token,
"sha256_refresh_token": sha256_refresh_token,"
> 
> When would the provided token be different than the hashed access/refresh token?

@misha-sentry The provided token will always match at least and at most one of these.

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes
